dsp/src/lockin.rs

```diff
@@ -135,120 +129,89 @@ impl Lockin {
     /// # Arguments
     ///
     /// * `adc_samples` - One batch of ADC samples.
-    /// * `timestamps` - Counter values corresponding to the edges of
-    /// an external reference signal. The counter is incremented by a
-    /// fast internal clock.
+    /// * `timestamp` - Counter value corresponding to the edges of an
+    /// external reference signal. The counter is incremented by a
+    /// fast internal clock. Each ADC sample batch can contain 0 or 1
+    /// timestamps.
     ///
     /// # Returns
     ///
     /// The demodulated complex signal as a `Result`. When there are
     /// an insufficient number of timestamps to perform processing,
     /// `Err` is returned.
-    ///
-    /// # Assumptions
-    ///
-    /// `demodulate` expects that the timestamp counter value is equal
-    /// to 0 when the ADC samples its first input in a batch. This can
-    /// be achieved by configuring the timestamp counter to overflow
-    /// at the end of the ADC batch sampling period.
     pub fn demodulate(
```

Comment:
I don't understand why this is still called `demodulate`. It does so much more.

* There seems to be the need for a bit of code (either a struct or just a `fn`) that does the timestamp handling (the branching, the division) before the PLL. Is it better to wrap the PLL to encapsulate all that (`update(Option<timestamp>) -> `(demodulation_frequency, demodulation_phase)`)?
* Then there is some code that does the sample processing. It seems prudent to not allocate and generate the complete demodulation signal and then do the mixing, and filtering, each with a batched loop with intermediate buffers. But instead move the code up into a single loop, if necessary within `process()`. The API is tiny now and it seems it doesn't need more wrapping. Also in the end the different use cases will look different. There may be some that will decimate, others may not. Yet others might have two biquads in chain. And there might be the `Unwrapper` as well. Some might take the phase, others might be interested only in the amplitude.

Maybe a good suggestion to understand what I mean would be for you to try to actually use the code in the processing ISR in `main.rs`. That should show whether these wrappers are actually convenient or whether the code would be much shorter and easier to understand if they were inlined.

Reply:
I've removed the lockin module completely and moved this code into main. A new `TimestampHandler` receives timestamps, wraps the PLL and computes the demodulation frequency and initial phase. Everything else is done in `process()`. Let me know if this is more along the lines of what you had in mind.